Add Manga Migration Feature + Library Sorting Improvements

Luskebusk opened the pull request:
## Summary
This PR implements a complete manga migration system allowing users to migrate manga between sources while preserving reading progress and categories. Also includes library sorting improvements for better manga organization.

## Features Added

### Manga Migration System
Complete migration workflow accessible from manga details screen:
- **Migration Entry Point**: Added "Migrate" option to manga details 3-dot menu
- **Global Search Integration**: Search for target manga across all available sources
- **Migration Preview**: Compare source and target manga with configurable options
- **Progress Tracking**: Real-time migration progress with detailed status updates
- **Data Preservation**: Migrates reading progress and categories

### Library Sorting Improvements
Enhanced library organization capabilities:
- **Last Updated Sorting**: Sort manga by most recent chapter fetch date
- **Persistent Sorting**: Sorting preferences saved across app sessions
- **Direction Toggle**: Ascending/descending sort with proper indicators

### ICU Formatting Fixes
- **Tamil Language**: Fixed ICU syntax errors preventing localization generation
- **Dependencies**: Updated `intl` dependency to `^0.20.2`

## Implementation Details

### Migration Architecture

lib/src/features/migration/
├── controller/          # Riverpod state management
├── data/               # Repository and GraphQL integration
├── domain/             # Models and data structures
└── presentation/       # UI screens and widgets

### Key Components
- **GraphQL Integration**: Uses existing Suwayomi server APIs for real migrations
- **Type Safety**: Full type safety with Freezed models and generated code
- **Error Handling**: Comprehensive error handling with user-friendly feedback
- **Cache Management**: Proper cache invalidation for immediate UI updates
- **Hidden non-functional migration options** (Downloads, Tracking) until implementation

## Files Changed

### Migration Feature (New Files)
- `lib/src/features/migration/controller/migration_controller.dart`
- `lib/src/features/migration/data/migration_repository.dart`
- `lib/src/features/migration/domain/migration_models.dart`
- `lib/src/features/migration/presentation/screens/migration_global_search_screen.dart`
- `lib/src/features/migration/presentation/screens/migration_preview_screen.dart`
- `lib/src/features/migration/presentation/screens/migration_progress_screen.dart`
- `lib/src/features/migration/presentation/widgets/migration_source_short_search.dart`

### Modified Files
- `lib/src/features/manga_book/presentation/manga_details/manga_details_screen.dart` - Added migration menu option
- `lib/src/constants/enum.dart` - Added `lastUpdated` sort option
- `lib/src/features/library/presentation/library/controller/library_controller.dart` - Enhanced sorting logic
- `lib/src/l10n/app_en.arb` - Added localization strings
- `lib/src/l10n/app_ta.arb` - Fixed ICU syntax errors
- `pubspec.yaml` - Updated intl dependency

## Migration Workflow

1. **Entry**: User taps 3-dot menu in manga details → "Migrate"
2. **Search**: Global search across all sources for target manga
3. **Selection**: Choose target manga from search results
4. **Preview**: Compare source vs target with migration options
5. **Execution**: Real-time progress tracking through migration stages
6. **Completion**: Automatic navigation back to library with refreshed data

## Migration Capabilities

### Data Migration
- **Reading Progress**: Chapter read status and last page read
- **Categories**: All assigned categories transferred
- **Downloads**: Hidden until implementation
- **Tracking**: Hidden until implementation

### Smart Features
- **Chapter Matching**: Intelligent matching by chapter number
- **Conflict Resolution**: Handles numbering differences between sources
- **Progress Tracking**: Real-time updates with cancellation support
- **Error Recovery**: Graceful handling of partial migrations

## Testing

### Functional Testing
- Complete migration flow from manga details to completion
- Global search integration with streaming results
- Migration preview with options configuration
- Progress tracking with real GraphQL operations
- Library sorting with persistence
- Localization generation without errors

### Data Integrity
- Reading progress preserved across migration
- Categories properly transferred
- Cache invalidation for immediate UI updates
- Source manga removal (when enabled)

## Dependencies
- Updated `intl` from `^0.19.0` to `^0.20.2`
- Uses existing GraphQL, Riverpod, and Freezed dependencies

## Breaking Changes
None. All changes are additive or fix existing functionality.

## Screenshots

### Migration Flow
1. **Migration Entry Point**: ![image](https://github.com/user-attachments/assets/d1e7b8c5-9185-4acf-9afb-0e1320112530)
2. **Global Search**: ![image](https://github.com/user-attachments/assets/f2a90200-6842-4e84-96f0-ad3865549f51)
3. **Migration Preview**: ![image](https://github.com/user-attachments/assets/3b0a94a9-ee09-42fa-85ec-0fb2a2f5d8d6)
4. **Progress Tracking**: ![image](https://github.com/user-attachments/assets/d66da543-9844-4c19-a6cf-9bcae99251b2)
5. **Library Sorting**: ![image](https://github.com/user-attachments/assets/6f0cb4ac-a65c-4840-9537-7a3828d0b33e)


### Key Features
- **Migration Options**: Configurable migration settings (chapters, categories)
- **Source Comparison**: Clear before/after manga information display
- **Progress Indicators**: Visual progress bars and status messages

DattatreyaReddy commented on lib/src/features/migration/presentation/screens/migration_search_screen.dart:
```diff
@@ -0,0 +1,330 @@
+    ref.read(selectedTargetMangaProvider.notifier).select(targetManga);
+    
+    // Navigate to migration preview screen
+    context.push(
```
Also use .go() instead of .push() if possible
- if it is not possible then thats ok

DattatreyaReddy commented on lib/src/features/migration/presentation/widgets/migration_source_short_search.dart:
```diff
@@ -0,0 +1,86 @@
+  
+  final SourceDto source;
+  final AsyncValue<List<MangaDto>> mangaList;
+  final dynamic sourceManga;
```
dont use dynamic

DattatreyaReddy commented on lib/src/routes/sub_routes/migration_routes.dart:
```diff
@@ -0,0 +1,139 @@
+
+  @override
+  Widget build(BuildContext context, GoRouterState state) {
+    final extra = state.extra as Map<String, dynamic>?;
```
create a new class or record and use that instead of Map

DattatreyaReddy reviewed (changes requested):
I thank you for contributing to Sorayomi.

I've pointed out few feedback points that need to be addressed. I'll try to merge this after reviewing again. You can expect the next release within a month,

Luskebusk commented:
@DattatreyaReddy  Thanks for the review! I addressed the issues you specified:

- Fixed dynamic types → proper typing throughout
- Replaced Theme.of(context) with context.theme  
- Added AsyncValue.guard for async error handling
- Split large widgets into smaller components
- Updated localization strings

For the navigation routes I've added proper typed data classes which achieves the type safety you mentioned. The .go() method doesn't work well with the code and it would require a lot of changes. If it's not a big problem or you can see another way to implement it I'd be happy to give it a try

Thanks again, this is my first time collaborating on git so thank you for your patience

DattatreyaReddy commented:
> @DattatreyaReddy Thanks for the review! I addressed the issues you specified:
> 
> * Fixed dynamic types → proper typing throughout
> * Replaced Theme.of(context) with context.theme
> * Added AsyncValue.guard for async error handling
> * Split large widgets into smaller components
> * Updated localization strings
> 
> For the navigation routes I've added proper typed data classes which achieves the type safety you mentioned. The .go() method doesn't work well with the code and it would require a lot of changes. If it's not a big problem or you can see another way to implement it I'd be happy to give it a try
> 
> Thanks again, this is my first time collaborating on git so thank you for your patience

Use MangaDetailsScreen().push(context) instead then (dont use context.push)

DattatreyaReddy commented:
by using $extra like this you won't neet to cast the extra

DattatreyaReddy commented:
Also run `git config --local core.hooksPath .githooks` to config pre commit hooks

DattatreyaReddy commented on lib/src/features/migration/controller/migration_controller.dart:
```diff
@@ -0,0 +1,320 @@
+
+      // Update progress to finalizing
+      state = const MigrationProgress(
+        currentStep: 'migrationInProgress',
```
You can also make this enum

DattatreyaReddy commented on lib/src/features/migration/presentation/widgets/migration_progress_widgets.dart:
```diff
@@ -0,0 +1,539 @@
+  }
+
+  IconData _getStepIcon(MigrationStatus status) {
+    switch (status) {
```
use dart switch statements instead of switch cases

DattatreyaReddy reviewed (changes requested):
You can take your time fixing this,

Luskebusk commented:
I see, thank you! I'll read through the link you sent and take my time with this. Thanks again!

DattatreyaReddy reviewed (approved):
Thank you for your contribution.
